packages/core/src/codecatalyst/auth.ts

```diff
@@ -118,6 +155,54 @@ export class CodeCatalystAuthenticationProvider {
         await this.secondaryAuth.restoreConnection()
     }
 
+    private async accessDeniedExceptionHandler() {
+        if (!this.isConnectionValid()) return
+
+        // Check if CodeWhisper and CodeCatalyst share the same connection
+        const isSharedConn =
+            this.secondaryAuth.activeConnection?.id == CodeWhispererAuth.instance.secondaryAuth.activeConnection?.id
+
+        if (isSharedConn) {
+            await codeWhispererClient.listAvailableCustomizations()
+        }
+
+        this.updateConnectionState(this.activeConnection!, { shared: isSharedConn })
+
+        const isPartialExpiration = isSharedConn && !CodeWhispererAuth.instance.isConnectionExpired()
+
+        getLogger().info('CodeCatalyst scopes are expired.')
+
+        this.setScopeExpired(true)
+        await setCodeCatalystConnectedContext(this.isConnectionValid())
+
+        this.showReauthenticationPrompt(this.activeConnection!, isPartialExpiration)
+    }
+
+    public async showReauthenticationPrompt(conn: Connection, isPartialExpiration?: boolean): Promise<void> {
+        const settings = PromptSettings.instance
+        const shouldShow = await settings.isPromptEnabled('codeCatalystConnectionExpired')
+        if (this.suppressNextReauthPrompt) {
+            this.suppressNextReauthPrompt = false
+            return
+        }
+        if (!shouldShow) {
+            return
+        }
+        const expiredMessage =
+            'Connection expired. To continue using CodeCatalyst, connect with AWS Builder ID or AWS IAM Identity center.'
+        const partiallyExpiredMessage =
+            'CodeCatalyst connection has expired. Amazon Q/CodeWhisperer is still connected.'
+        const message = isPartialExpiration ? partiallyExpiredMessage : expiredMessage
+        const connect = 'Connect with AWS'
+        const doNotShow = "Don't Show Again"
+        const resp = await vscode.window.showInformationMessage(message, connect, doNotShow)
+        if (resp === connect) {
+            await this.auth.reauthenticate(conn)
+        } else if (resp === doNotShow) {
+            await settings.disablePrompt('codeCatalystConnectionExpired')
+        }
```

Comment:
This looks like a copy-paste from https://github.com/aws/aws-toolkit-vscode/blob/e5fdf4109207f801a9b91a4d162ef1fb8d1bffd1/packages/core/src/codewhisperer/util/authUtil.ts#L358-L370

can we abstract it to a common function in [messages.ts](https://github.com/aws/aws-toolkit-vscode/blob/e5fdf4109207f801a9b91a4d162ef1fb8d1bffd1/packages/core/src/shared/utilities/messages.ts)

Reply:
Should be done in this PR. Copy-paste is not an acceptable practice and the expected result is that this PR will be smaller, since factoring out common code reduces duplication.

Reply:
Added to this PR